test/infrastructure/docker/internal/docker/machine.go

```diff
@@ -419,6 +423,71 @@ func (m *Machine) SetNodeProviderID(ctx context.Context, c client.Client) error
 	return nil
 }
 
+// CloudProviderNodePatch sets the docker provider ID for the kubernetes node.
+func (m *Machine) CloudProviderNodePatch(ctx context.Context, c client.Client, dockerMachine *infrav1.DockerMachine) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	kubectlNode, err := m.getKubectlNode(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "unable to set NodeProviderID. error getting a kubectl node")
+	}
+	if !kubectlNode.IsRunning() {
+		return errors.Wrapf(ContainerNotRunningError{Name: kubectlNode.Name}, "unable to set NodeProviderID")
+	}
+
+	node := &corev1.Node{}
+	if err = c.Get(ctx, apimachinerytypes.NamespacedName{Name: m.ContainerName()}, node); err != nil {
+		return errors.Wrap(err, "failed to retrieve node")
+	}
+
+	patchHelper, err := patch.NewHelper(node, c)
+	if err != nil {
+		return err
+	}
+
+	// If the ProviderID is set ensure the cloudprovider taint has been removed.
+	if dockerMachine.Spec.ProviderID != nil {
```

Comment:
Given that we know that this didn't help I assume we can use a more logical order and do this at the end of the func together with setting the providerID and the addresses?